async_server.c

```diff
@@ -0,0 +1,381 @@
+
+cleanup:
+  if (queue_cancel (&ctx->registry_idx) != QS_SUCCESS)
+    error ("Could not cancel registry indices queue");
+
+  if (queue_destroy (&ctx->registry_idx) != QS_SUCCESS)
+    error ("Could not destroy registry indices queue");
+
+  free (ctx);
+
+  return (NULL);
+}
+
+static void
+acl_context_destroy (acl_context_t *ctx)
+{
+  trace ("Destroying client context");
+
+  if (queue_cancel (&ctx->registry_idx) != QS_SUCCESS)
+    error ("Could not cancel registry indices queue");
+
+  trace ("Cancelled registry indices queue");
+
+  if (queue_destroy (&ctx->registry_idx) != QS_SUCCESS)
+    error ("Could not destroy registry indices queue");
+
+  trace ("Destroyed registry indices queue");
+
+  if (pthread_mutex_destroy (&ctx->mutex) != 0)
+    error ("Could not destroy mutex");
+
+  close_client (ctx->socket_fd);
+  free (ctx);
+}
+
+static status_t
+return_tasks (acl_context_t *ctx)
+{
+  mt_context_t *mt_ctx = &ctx->context->context;
+  status_t status = S_SUCCESS;
+
+  int i;
+  for (i = 0; i < QUEUE_SIZE; ++i)
+    {
+      if (ctx->registry_used[i])
+        {
+          if (queue_push_back (&mt_ctx->queue, &ctx->registry[i])
+              != QS_SUCCESS)
+            {
+              status = S_FAILURE;
+              break;
+            }
+          ctx->registry_used[i] = false;
+        }
+    }
+
+  return (status);
+}
+
+static void
+sender_receiver_cleanup (void *arg)
+{
+  acl_context_t *ctx = arg;
+
+  bool is_last;
+  if (pthread_mutex_lock (&ctx->mutex) != 0)
+    return;
+  pthread_cleanup_push (cleanup_mutex_unlock, &ctx->mutex);
+
+  return_tasks (ctx);
+  is_last = (--ctx->ref_count == 0);
+
+  pthread_cleanup_pop (!0);
+
+  if (is_last)
+    acl_context_destroy (ctx);
+}
+
+static void *
+result_receiver (void *arg)
+{
+  acl_context_t *cl_ctx = *(acl_context_t **)arg;
+  mt_context_t *mt_ctx = &cl_ctx->context->context;
+
+  pthread_cleanup_push (sender_receiver_cleanup, cl_ctx);
+  while (true)
+    {
+      result_t task;
+      if (recv_wrapper (cl_ctx->socket_fd, &task, sizeof (task), 0)
+          == S_FAILURE)
+        {
+          error ("Could not receive result from client");
+          break;
+        }
+
+      if (task.is_correct)
+        {
+          if (queue_cancel (&mt_ctx->queue) == QS_FAILURE)
+            {
+              error ("Could not cancel a queue");
+              break;
+            }
+          memcpy (mt_ctx->password, task.password, sizeof (task.password));
+        }
+
+      trace ("Received %s result %s from client",
+             task.is_correct ? "correct" : "incorrect", task.password);
+
+      if (queue_push (&cl_ctx->registry_idx, &task.id) != QS_SUCCESS)
+        {
+          error ("Could not return id to a queue");
+          break;
+        }
+
+      trace ("Pushed index of received task back to indices queue");
+
+      cl_ctx->registry_used[task.id] = false;
+
+      if (serv_signal_if_found (mt_ctx) == S_FAILURE)
+        break;
+
+      if (mt_ctx->password[0] != 0)
```

Comment:
Is it really required?

Reply:
Somewhy I thought that it is but its not required